cpp/src/io/comp/uncomp.cpp

```diff
@@ -498,14 +501,50 @@ size_t decompress_snappy(host_span<uint8_t const> src, host_span<uint8_t> dst)
   return uncompressed_size;
 }
 
+/**
+ * @brief ZSTD decompressor that uses nvcomp
+ */
+size_t decompress_zstd(host_span<uint8_t const> src,
+                       host_span<uint8_t> dst,
+                       rmm::cuda_stream_view stream)
+{
+  // Init device span of spans (source)
+  auto const d_src = cudf::detail::make_device_uvector_async(src, stream);
+  auto hd_srcs     = hostdevice_vector<device_span<uint8_t const>>(1, stream);
+  hd_srcs[0]       = d_src;
+  hd_srcs.host_to_device(stream);
+
+  // Init device span of spans (temporary destination)
+  auto d_dst   = rmm::device_uvector<uint8_t>(dst.size(), stream);
+  auto hd_dsts = hostdevice_vector<device_span<uint8_t>>(1, stream);
+  hd_dsts[0]   = d_dst;
+  hd_dsts.host_to_device(stream);
+
+  auto hd_stats                   = hostdevice_vector<decompress_status>(1, stream);
+  auto const max_uncomp_page_size = dst.size();
+  nvcomp::batched_decompress(
+    nvcomp::compression_type::ZSTD, hd_srcs, hd_dsts, hd_stats, max_uncomp_page_size, stream);
+
+  // Copy temporary output to `dst`
+  CUDF_CUDA_TRY(
+    cudaMemcpyAsync(dst.data(), d_dst.data(), dst.size(), cudaMemcpyDeviceToHost, stream.value()));
```

Comment:
```suggestion
    cudaMemcpyAsync(dst.data(), d_dst.data(), max_uncomp_page_size, cudaMemcpyDeviceToHost, stream.value()));
```

Reply:
just realized from this suggestion that we don't need to copy the whole buffer, just the portion that contains the actual output.